File: rust-api/frontendmentor/src/models/challenge_model.rs
use super::gallery_model::Gallery;
use serde::Deserialize;
use serde_derive::Serialize;

#[derive(Debug, Serialize)]
pub enum ChallengeType {
    #[serde(rename = "free")]
    Free,
    #[serde(rename = "free+")]
    FreePlus,
    #[serde(rename = "premium")]
    Premium,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Difficulty {
    #[serde(rename = "newbie")]
    Newbie,
    #[serde(rename = "junior")]
    Junior,
    #[serde(rename = "intermediate")]
    Intermediate,
    #[serde(rename = "advanced")]
    Advanced,
    #[serde(rename = "guru")]
    Guru,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Language {
    #[serde(rename = "html")]
    Html,
    #[serde(rename = "css")]
    Css,
    #[serde(rename = "js")]
    Js,
    #[serde(rename = "api")]
    Api,
    #[serde(rename = "general")]
    General,
}

impl ToString for Language {
    fn to_string(&self) -> String {
        match self {
            Language::Html => "html".to_owned(),
            Language::Css => "css".to_owned(),
            Language::Js => "js".to_owned(),
            Language::Api => "api".to_owned(),
            Language::General => "general".to_owned(),
        }
    }
}

#[derive(Debug, Serialize)]
pub struct Challenge {
    pub challenge_id: i32,

    // #[serde(rename(serialize = "createdAt"))]
    // pub created_at: NaiveDateTime,

    // #[serde(rename(serialize = "updatedAt"))]
    // pub updated_at: NaiveDateTime,
    #[serde(rename = "type")]
    pub challenge_type: String,

    pub title: String,

    pub difficulty: String,

    #[serde(rename(serialize = "startedCount"))]
    pub started_count: i32,

    #[serde(rename(serialize = "completedCount"))]
    pub completed_count: i32,

    pub description: String,

    // pub languages: Vec<String>,
    #[serde(rename(serialize = "heroImage"))]
    pub hero_image: String,

    pub brief: String,

    pub gallery: Vec<Gallery>,

    #[serde(rename(serialize = "starter_url"))]
    pub starter_url: String,
}

File: client/projects/apps/frontendmentor/src/app/shared/models/challenge.model.ts
type Gallery = ChallengeImage[];

export interface ChallengeImage {
  preview: string;
  design: string;
  title: string;
}
/**
 * - Challenge Query
 */
export interface ChallengeQuery {
  type: ChallengeType[] | undefined;
  difficulty: Difficulty[] | undefined;
  language: Language[] | undefined;
  sort: ChallengeSort | undefined;
}

/**
 * - Difficulty
 */
export type Difficulty =
  | 'newbie'
  | 'junior'
  | 'intermediate'
  | 'advanced'
  | 'guru';

export const DIFFICULTIES = {
  newbie: 1,
  junior: 2,
  intermediate: 3,
  advanced: 4,
  guru: 5,
};

/**
 * - Language
 */
export type Language = 'html' | 'css' | 'js' | 'api' | 'general';

/**
 * - Challenge Type
 */
export type ChallengeType = 'free' | 'free+' | 'premium';

/**
 * - Challenge Sort
 */
export enum ChallengeSort {
  DIFFICULTY_DESC = 'difficulty|desc',
  DIFFICULTY_ASC = 'difficulty|asc',
}

export type ChallengeStatus = 'in-progress' | 'completed';

/**
 * - Challenge
 */
export interface Challenge {
  id: string;
  title: string;
  difficulty: Difficulty;
  description: string;
  languages: Language[];
  heroImage: string;
  isNew?: boolean;
  type: ChallengeType;
  gallery: Gallery;
  status?: ChallengeStatus;
  staterURL: string;
  startedCount: number;
  completedCount: number;
  ideas: string;
  brief: string;
  likes: number;
  bookmarks: number;
  comments: number;
}

/**
 * - Identify Challenge
 * @param index
 * @param challenge
 * @returns
 */
export const identifyChallenge = (index: number, challenge: Challenge) => {
  return challenge.title;
};
